moshi/src/main/java/com/squareup/moshi/JsonWriter.java

```diff
@@ -483,165 +247,17 @@ public JsonWriter value(long value) throws IOException {
-      String replacement;
-      if (c < 128) {
-        replacement = replacements[c];
-        if (replacement == null) {
-          continue;
-        }
-      } else if (c == '\u2028') {
-        replacement = "\\u2028";
-      } else if (c == '\u2029') {
-        replacement = "\\u2029";
-      } else {
-        continue;
-      }
-      if (last < i) {
-        sink.writeUtf8(value, last, i);
-      }
-      sink.writeUtf8(replacement);
-      last = i + 1;
-    }
-    if (last < length) {
-      sink.writeUtf8(value, last, length);
-    }
-    sink.writeByte('"');
-  }
-
-  private void newline() throws IOException {
-    if (indent == null) {
-      return;
-    }
-
-    sink.writeByte('\n');
-    for (int i = 1, size = stackSize; i < size; i++) {
-      sink.writeUtf8(indent);
-    }
-  }
-
-  /**
-   * Inserts any necessary separators and whitespace before a name. Also
-   * adjusts the stack to expect the name's value.
-   */
-  private void beforeName() throws IOException {
-    int context = peek();
-    if (context == NONEMPTY_OBJECT) { // first in object
-      sink.writeByte(',');
-    } else if (context != EMPTY_OBJECT) { // not in an object!
-      throw new IllegalStateException("Nesting problem.");
-    }
-    newline();
-    replaceTop(DANGLING_NAME);
-  }
-
-  /**
-   * Inserts any necessary separators and whitespace before a literal value,
-   * inline array, or inline object. Also adjusts the stack to expect either a
-   * closing bracket or another element.
-   */
-  @SuppressWarnings("fallthrough")
-  private void beforeValue() throws IOException {
-    switch (peek()) {
-      case NONEMPTY_DOCUMENT:
-        if (!lenient) {
-          throw new IllegalStateException(
-              "JSON must have only one top-level value.");
-        }
-        // fall-through
-      case EMPTY_DOCUMENT: // first in document
-        replaceTop(NONEMPTY_DOCUMENT);
-        break;
-
-      case EMPTY_ARRAY: // first in array
-        replaceTop(NONEMPTY_ARRAY);
-        newline();
-        break;
-
-      case NONEMPTY_ARRAY: // another in array
-        sink.writeByte(',');
-        newline();
-        break;
-
-      case DANGLING_NAME: // value for name
-        sink.writeUtf8(separator);
-        replaceTop(NONEMPTY_OBJECT);
-        break;
-
-      default:
-        throw new IllegalStateException("Nesting problem.");
-    }
-  }
+  public abstract JsonWriter value(Number value) throws IOException;
 
   /**
    * Changes the reader to treat the next string value as a name. This is useful for map adapters so
    * that arbitrary type adapters can use {@link #value(String)} to write a name value.
    */
-  void promoteNameToValue() throws IOException {
-    int context = peek();
-    if (context != NONEMPTY_OBJECT && context != EMPTY_OBJECT) {
-      throw new IllegalStateException("Nesting problem.");
-    }
-    promoteNameToValue = true;
-  }
+  abstract void promoteNameToValue() throws IOException;
```

Comment:
This package scoped method (and equivalent in `JsonReader`) prevent these types from being fully re-implementable by consumers. We could switch to having it be public.